mineral-ui/mineral-ui
# Render props

### Description

Refactors **Select/Dropdown/Popover/Tooltip/Menu/MenuItem** components to add support for [render props](https://reactjs.org/docs/render-props.html), thus giving users greater custom rendering control.

### Motivation and context

* Provide greater custom rendering control to consumers.
* Make mineral components more flexible for a larger variety of use cases.

closes #628 

_Differences from original AC:_

* Decided to use prop objects rather than prop getters as it was easier to understand and easier to compose.
* Decided not to add extra highlightedIndex functionality in Select/Dropdown at this time because we could not think of an actual use case.  We will add it if it is requested.

### Screenshots, videos, or demo, if appropriate

https://628-render-props--mineral-ui.netlify.com/

### How to test

* Check out the aforementioned components in the [demo](https://628-render-props--mineral-ui.netlify.com/).  Each one of them has new "Custom" examples.
* Note that VR tests are mostly useless here since they don't run on components using Popper.  The diffs that do exists look to be expected.

### Types of changes
- New feature (non-breaking change which adds functionality)
  * New props
- Breaking change (fix or feature that would cause existing functionality to change)
  * Select, Dropdown, Menu no longer support `item.render` in data.  Use `renderItem` prop instead.

### Checklist

* [x] Rendering performance (initial load time & 60fps) and [perceived performance](http://blog.teamtreehouse.com/perceived-performance) considered. - These components likely do unnecessary re-rendering and thus performance can be improved.  Performance optimization is however complex and time consuming and thus needs to be done as a separate effort.  Suggestions in this area would be very welcomed.
